media/multipart/common/src/main/java/io/helidon/media/multipart/common/MultiPartDecoder.java

```diff
@@ -0,0 +1,310 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.media.multipart.common;
+
+import java.util.LinkedList;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.Flow.Processor;
+import java.util.concurrent.Flow.Subscriber;
+import java.util.concurrent.Flow.Subscription;
+import java.util.logging.Logger;
+
+import io.helidon.common.http.DataChunk;
+import io.helidon.common.reactive.OriginThreadPublisher;
+import io.helidon.media.common.MessageBodyReadableContent;
+import io.helidon.media.common.MessageBodyReaderContext;
+
+/**
+ * Reactive processor that decodes HTTP payload as a stream of {@link BodyPart}.
+ * This processor is a single use publisher that supports a single subscriber,
+ * it is not resumable.
+ */
+public final class MultiPartDecoder implements Processor<DataChunk, ReadableBodyPart> {
+
+    /**
+     * Logger.
+     */
+    private static final Logger LOGGER = Logger.getLogger(MultiPartDecoder.class.getName());
+
+    /**
+     * Indicate that the chunks subscription is complete.
+     */
+    private boolean complete;
+
+    /**
+     * The upstream subscription.
+     */
+    private Subscription chunksSubscription;
+
+    /**
+     * The builder for the current {@link BodyPart}.
+     */
+    private ReadableBodyPart.Builder bodyPartBuilder;
+
+    /**
+     * The builder for the current {@link ReadableBodyPartHeaders}.
+     */
+    private ReadableBodyPartHeaders.Builder bodyPartHeaderBuilder;
+
+    /**
+     * The publisher for the current part.
+     */
+    private BodyPartContentPublisher contentPublisher;
+
+    /**
+     * The parent chunk for the published body part chunks.
+     */
+    private BodyPartChunk.Parent chunkParent;
+
+    /**
+     * The MIME parser.
+     */
+    private final MIMEParser parser;
+
+    /**
+     * The parser event processor.
+     */
+    private final ParserEventProcessor parserEventProcessor;
+
+    /**
+     * The bodyParts processed during each {@code onNext}.
+     */
+    private final Queue<ReadableBodyPart> bodyParts;
+
+    /**
+     * The reader context.
+     */
+    private final MessageBodyReaderContext context;
+
+    /**
+     * The downstream publisher.
+     */
+    private final BodyPartPublisher downstream;
+
+    /**
+     * Create a new instance.
+     *
+     * @param boundary mime message boundary
+     * @param context reader context
+     */
+    private MultiPartDecoder(String boundary, MessageBodyReaderContext context) {
+        Objects.requireNonNull(boundary, "boundary cannot be null!");
+        Objects.requireNonNull(context, "context cannot be null!");
+        this.context = context;
+        parserEventProcessor = new ParserEventProcessor();
+        parser = new MIMEParser(boundary, parserEventProcessor);
+        downstream = new BodyPartPublisher();
```

Comment:
We have nice tooling for that, no need to misuse OriginThreadPublisher here 
[MultiPartDecoder_not_extending_OTPub.patch.txt](https://github.com/oracle/helidon/files/4627871/MultiPartDecoder_not_extending_OTPub.patch.txt)

Reply:
I wrote most of this code prior to all of the new Multi stuff ; which is it is using OTP. I'll take a look at your patch.

Reply:
Looking at your diff, there seems to be a part that's missing. Let's schedule something next week to go over your suggestions.